ParProcCo/job_controller.py

```diff
@@ -5,91 +5,83 @@
 from pathlib import Path
 from typing import List, Optional
 
-from ParProcCo.scheduler_mode_interface import SchedulerModeInterface
-from ParProcCo.job_scheduler import JobScheduler
-from ParProcCo.slicer_interface import SlicerInterface
-from ParProcCo.utils import check_location, get_absolute_path
-
+from .job_scheduler import JobScheduler
+from .slicer_interface import SlicerInterface
+from .utils import check_location, get_absolute_path
+from .program_wrapper import ProgramWrapper
 
 class JobController:
 
-    def __init__(self, cluster_output_dir_name: str, project: str, queue: str, cluster_resources: Optional[dict[str,str]] = None, timeout: timedelta = timedelta(hours=2)):
+    def __init__(self, program_wrapper: ProgramWrapper, cluster_output_dir_name: str, project: str, queue: str, cluster_resources: Optional[dict[str,str]] = None, timeout: timedelta = timedelta(hours=2)):
         """JobController is used to coordinate cluster job submissions with JobScheduler"""
-
+        self.program_wrapper = program_wrapper
         self.working_directory: Path = check_location(os.getcwd())
-        self.aggregation_scheduler: JobScheduler
         self.cluster_output_dir: Path = self.working_directory / cluster_output_dir_name
         self.data_slicer: SlicerInterface
         self.project = project
         self.queue = queue
         self.cluster_resources = cluster_resources
-        self.job_scheduler: JobScheduler
         self.timeout = timeout
-        self.scheduler: JobScheduler
+        self.sliced_results: Optional[List[Path]] = None
 
-    def run(self, data_slicer: SlicerInterface, number_jobs: int, processing_mode: SchedulerModeInterface,
-            aggregating_mode: SchedulerModeInterface, cluster_runner: Path,
-            jobscript_args: Optional[List] = None, aggregator_path: Optional[Path] = None, memory: str = "4G", cores: int = 6,
-            job_name: str = "ParProcCo") -> None:
+    def run(self, number_jobs: int, cluster_runner: Path,
+            jobscript_args: Optional[List] = None, aggregator_path: Optional[Path] = None,
+            memory: str = "4G", cores: int = 6, job_name: str = "ParProcCo") -> None:
 
-        slice_params = self.create_slices(data_slicer, number_jobs)
+        slice_params = self.program_wrapper.create_slices(number_jobs) if number_jobs > 1 else []
```

Comment:
I do not think setting this to `[]` works:
`len(slice_params)` is used to set number of jobs in `SimpleProcessingMode` which is used in L94 of the `JobScheduler.run` method:
`job_indices = list(range(scheduler_mode.number_jobs))`. If `job_indices` is empty no job template are created and run.

Reply:
Needs to use PassThruMode

Reply:
Yes, but `msm_cluster_submit` sets the wrapper to be `PassThruMode` when `jobs==1`

Reply:
in test_job_controller, `test_single_job_does_not_aggregate` doesn't pass.

Reply:
I thought I tested it!

Reply:
Missed commit my fix!